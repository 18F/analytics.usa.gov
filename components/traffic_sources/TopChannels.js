import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../js/lib/renderblock";

function TopChannels({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/top-session-channel-group-30-days.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initTopChannelsChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock.buildCompactBarChart("session_default_channel_group"),
        );
      return result;
    };
    initTopChannelsChart().catch(console.error);
  }, []);

  return (
    <figure id="chart_session_channel_group" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopChannels.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopChannels;
