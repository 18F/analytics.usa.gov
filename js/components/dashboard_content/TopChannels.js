import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";

/**
 * Retrieves the top session channel group report from the passed data URL and
 * creates a visualization for the count of users visiting sites for the current
 * agency with each channel.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopChannels({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/top-session-channel-group-30-days.json`;
  const ref = useRef(null);
  const [channelData, setChannelData] = useState(null);

  useEffect(() => {
    const initChannelsChart = async () => {
      if (!channelData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setChannelData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder.buildCompactBarChart(
          ref.current,
          channelData,
          "session_default_channel_group",
        );
      }
    };
    initChannelsChart().catch(console.error);
  }, [channelData]);

  return (
    <figure id="chart_session_channel_group" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopChannels.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopChannels;
