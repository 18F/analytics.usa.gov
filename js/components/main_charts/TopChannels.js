import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";

/**
 * Retrieves the top session channel group report from the passed data URL and
 * creates a visualization for the count of users visiting sites for the current
 * agency with each channel.
 *
 * @param {String} dataHrefBase the URL of the base location of the data to be
 * downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 */
function TopChannels({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/top-session-channel-group-30-days.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initTopChannelsChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock.buildCompactBarChart("session_default_channel_group"),
        );
      return result;
    };
    initTopChannelsChart().catch(console.error);
  }, []);

  return (
    <figure id="chart_session_channel_group" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopChannels.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopChannels;
