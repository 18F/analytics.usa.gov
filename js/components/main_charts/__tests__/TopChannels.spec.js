import renderer from "react-test-renderer";
import TopChannels from "../TopChannels";

describe("TopChannels", () => {
  it("renders", () => {
    const component = renderer.create(
      <TopChannels dataHrefBase="http://www.example.com/data/" />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
