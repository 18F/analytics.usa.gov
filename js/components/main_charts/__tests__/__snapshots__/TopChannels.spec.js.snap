// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopChannels renders 1`] = `
<figure
  id="chart_session_channel_group"
>
  <div
    className="data bar-chart"
  />
</figure>
`;
